src/python/library/tritonclient/utils/shared_memory/__init__.py

```diff
@@ -146,41 +128,35 @@ def set_shared_memory_region(shm_handle, input_values, offset=0):
     """
 
     if not isinstance(input_values, (list, tuple)):
-        _raise_error("input_values must be specified as a list/tuple of numpy arrays")
+        raise SharedMemoryException(
+            "input_values must be specified as a list/tuple of numpy arrays"
+        )
     for input_value in input_values:
         if not isinstance(input_value, np.ndarray):
-            _raise_error("each element of input_values must be a numpy array")
+            raise SharedMemoryException(
+                "each element of input_values must be a numpy array"
+            )
 
-    offset_current = offset
-    for input_value in input_values:
-        input_value = np.ascontiguousarray(input_value).flatten()
-        if input_value.dtype == np.object_:
-            input_value = input_value.item()
-            byte_size = np.dtype(np.byte).itemsize * len(input_value)
-            _raise_if_error(
-                c_int(
-                    _cshm_shared_memory_region_set(
-                        shm_handle,
-                        c_uint64(offset_current),
-                        c_uint64(byte_size),
-                        cast(input_value, c_void_p),
-                    )
+    try:
+        for input_value in input_values:
+            # numpy array of object type is "syntactic sugar" for the API, should
+            # be handled by accessing its item and treat as Python object
+            if input_value.dtype == np.object_:
+                byte_size = len(input_value.item())
+                shm_handle._mpsm_handle.buf[offset : offset + byte_size] = (
+                    input_value.item()
                 )
-            )
-        else:
-            byte_size = input_value.size * input_value.itemsize
-            _raise_if_error(
-                c_int(
-                    _cshm_shared_memory_region_set(
-                        shm_handle,
-                        c_uint64(offset_current),
-                        c_uint64(byte_size),
-                        input_value.ctypes.data_as(c_void_p),
-                    )
+                offset += byte_size
+            else:
+                shm_tensor_view = np.ndarray(
+                    input_value.shape,
+                    input_value.dtype,
+                    buffer=shm_handle._mpsm_handle.buf[offset:],
                 )
-            )
-        offset_current += byte_size
-    return
+                shm_tensor_view[:] = input_value[:]
```

Comment:
Doesn't `buffer=shm_handle._mpsm_handle.buf[offset:],` fill the array? Are we immediately overwriting it with  `shm_tensor_view[:] = input_value[:]`?

https://numpy.org/devdocs/reference/generated/numpy.ndarray.html

> buffer : object exposing buffer interface, optional
>   **Used to fill the array with data.**

Reply:
The passed buffer is the underlying buffer of the numpy array, this method is used in the [shared memroy example](https://docs.python.org/3/library/multiprocessing.shared_memory.html#multiprocessing.shared_memory.SharedMemory.size)